src/common/transformations/src/transformations/mlir/mlir_op.cpp

```diff
@@ -267,6 +290,138 @@ namespace mlir {
 
 using namespace ::mlir;
 
+std::shared_ptr<MLIREvaluateBase> MLIREvaluateBase::create(OwningOpRef<ModuleOp> module,
+                                                           MlirMode mode,
+                                                           std::shared_ptr<ov::EvaluationContext> loweringContext) {
+    switch (mode) {
+        #if defined(GRAPH_COMPILER) && defined(GC_ENABLE_IMEX)
+        case MLIR_MODE_GC_GPU:
+            return std::make_shared<MLIREvaluateGcGPU>(std::move(module), loweringContext);
+        #endif
+        case MLIR_MODE_TPP:
+        case MLIR_MODE_GC:
+        case MLIR_MODE_DEFAULT:
+            return std::make_shared<MLIREvaluate>(std::move(module), mode);
+        default:
+            OPENVINO_THROW("Unsupported MLIR mode");
+    }
+}
+
+#if defined(GRAPH_COMPILER) && defined(GC_ENABLE_IMEX)
+
+cl_device_id extract_device_from_context(cl_context context) {
+    size_t devices_size;
+    cl_int err = clGetContextInfo(context, CL_CONTEXT_DEVICES, 0, NULL, &devices_size);
+    if (err != CL_SUCCESS) {
+        OPENVINO_THROW("Error getting context info: ", err);
+    }
+    if (devices_size / sizeof(cl_device_id) != 1) {
+        OPENVINO_THROW("Expected exactly one device in the context, got ", devices_size);
+    }
+
+    cl_device_id devices;
+    err = clGetContextInfo(context, CL_CONTEXT_DEVICES, devices_size, &devices, NULL);
+    if (err != CL_SUCCESS) {
+        OPENVINO_THROW("Error getting device IDs: ", err);
+    }
+
+    return devices;
+}
+
+
+
+MLIREvaluateGcGPU::MLIREvaluateGcGPU(OwningOpRef<mlir::ModuleOp> _module, std::shared_ptr<ov::EvaluationContext> loweringContext) {
+    OPENVINO_MLIR_DEBUG_PRINT(
+        "[ DEBUG ] Source MLIR:\n"
+        "-----------------------------------------\n");
+    OPENVINO_MLIR_DEBUG(_module->dump());
+    OPENVINO_MLIR_DEBUG_PRINT(
+        "-----------------------------------------\n");
+
+    gc::gpu::OclModuleBuilder builder(std::move(_module));
+
+    auto it = loweringContext->find(ov::intel_gpu::ocl_context.name());
+    if (it == loweringContext->end()) {
+        OPENVINO_THROW("No cl_context provided for OpenCL execution");
+    }
+    // QUESTION: are we sure that there's always only one device per context?
+    // QUESTION: may we pass context as void* directly to 'builder.build()' and make it responsible
+    // for extracting the device?
```

Comment:
If we don't know the target device at this point, then we should save the builder here and call build() in invoke().

Reply:
In fact we do know the target device. There are several options of how we can extract device info in [`transformations_pipeline.cpp`](https://github.com/dchigarev/openvino/blob/de413d45ab7088b4ae9cb8c80c1207904f17d5db/src/plugins/intel_gpu/src/plugin/transformations_pipeline.cpp#L851-L858):

<b>1. Extract an actual `cl_device_id`:</b>
```cpp
cl_device_id device = dynamic_cast<ocl_device>(m_context.get_engine().get_device()).get_device().get();
```
The problem with this approach is that we need to dynamic-cast an instance of an abstract `device` class to `ocl_device`. The definition of `ocl_device` is not available in `transformations_pipeline.cpp` without [this hack](https://github.com/slyalin/openvino/pull/169/files#r1780735084). Since I don't know yet what openvino guys think regarding this headers exposure, I tried not to rely on it where possible. But if they would say that it's fine we could easily forward a `cl_device_id` here is well.

<b>2. Extract device name as a string:</b>
```cpp
std::string device = m_context.get_device_name();
```
We could forward this device name to the `MLIREvaluateGcGPU` constructor and a device with the matching name if there are several devices in the context.

<b>3. Extract `cldnn::device_info`:</b>
```cpp
cldnn::device_info device = this->device_info;
```
The [device info structure](https://github.com/dchigarev/openvino/blob/dc78071ce15f9c481088a2bd42d4e49a89181d3e/src/plugins/intel_gpu/include/intel_gpu/runtime/device_info.hpp#L50) contains a lot of useful information that could also help us to identify which device we need to extract from this `cl_context`.

-----

The first options is the easiest one. So let's see what OV guys reply regarding the [headers exposure](https://github.com/slyalin/openvino/pull/169/files#r1780735084) and then decide how we want to proceed here.